File: Python/LeetCode/Num203.py
"""
Date:
  23.06.02
Number:
  LeetCode 203
Title:
  Remove Linked List Elements
Level:
  Easy
Author:
  thelight0804
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
  def removeElements(self, head, val):
    """
    :type head: ListNode
    :type val: int
    :rtype: ListNode
    """
    dummyNode = ListNode(-1) #create dummy node
    dummyNode.next = head #link nodes

    crntNode = head #current node
    prevNode = dummyNode #previous node

    while crntNode:
      if crntNode.val == val: #equal value
        prevNode.next = crntNode.next #link next node
        crntNode = crntNode.next #move crntNode
      else:
        crntNode = crntNode.next #move crntNode
        prevNode = prevNode.next #move prevNode
    
    return dummyNode.next
    

File: JavaScript/LeetCode/Num203.js
/**
 * Date: 23.09.13
 * Number: LeetCode 203
 * Title: Remove Linked List Elements
 * Level: Easy
 * Author: thelight0804
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function(head, val) {
    // create dummy node
    var dummyNode = new ListNode(-1);
    dummyNode.next = head;

    var crntNode = head; //current node
    var prevNode = dummyNode; // previous node

    while (crntNode){
        // remove element
        if (crntNode.val === val){
            prevNode.next = crntNode.next;
            crntNode = crntNode.next;
        }
        else { // move prevNode and crntNode
            prevNode = crntNode;
            crntNode = crntNode.next;
        }
    }

    return dummyNode.next;
};